dynamic_instrumentation/src/instrument.rs

```diff
@@ -531,49 +653,51 @@ impl<'tcx> Visitor<'tcx> for CollectInstrumentationPoints<'_, 'tcx> {
                 let source = remove_outer_deref(*p, self.tcx());
-                if let BorrowKind::Mut { .. } = bkind {
-                    // Instrument which local's address is taken
-                    self.loc(location, location, addr_local_fn)
-                        .arg_addr_of(*p)
-                        .arg_index_of(p.local)
-                        .source(&source)
-                        .dest(&dest)
-                        .instrumentation_priority(InstrumentationPriority::Early)
-                        .add_to(self);
-                } else {
-                    // Instrument immutable borrows by tracing the reference itself
-                    self.loc(location, location.successor_within_block(), addr_local_fn)
-                        .arg_var(dest)
-                        .arg_index_of(p.local)
-                        .source(&source)
+                let layout = ctx
+                    .layout_of(ty::ParamEnv::reveal_all().and(local_ty(p)))
+                    .expect("Failed to compute layout of local");
+                let size =
+                    u32::try_from(layout.size.bytes()).expect("Failed to convert local size");
+
+                let mut b = self
+                    .loc(location, location, addr_local_fn)
+                    .arg_addr_of(p.local)
+                    .arg_index_of(p.local)
+                    .arg_var(size)
+                    .source(&source)
+                    .instrumentation_priority(InstrumentationPriority::Early);
+                if p.projection.is_empty() {
+                    // No projections, store directly to the destination
+                    b = b.dest(&dest);
+                }
+                b.add_to(self);
+
+                if !p.projection.is_empty() {
+                    let fields = p
+                        .projection
+                        .iter()
+                        .filter_map(|elem| match elem {
+                            PlaceElem::Field(idx, _) => Some(idx.index()),
+                            _ => None,
+                        })
+                        .collect::<Vec<usize>>();
+                    let proj_idx = self.projections.add_proj(fields);
+
+                    let mut b = self
+                        .loc(location, location, project_fn)
+                        .arg_addr_of(p.local)
                         .dest(&dest)
-                        .instrumentation_priority(InstrumentationPriority::Early)
-                        .add_to(self);
-                };
+                        .instrumentation_priority(InstrumentationPriority::Early);
+                    if matches!(value, Rvalue::Ref(_, BorrowKind::Mut { .. }, _)) {
+                        b = b.arg_addr_of(*p);
+                    } else {
+                        b = b.arg_var(dest);
+                    }
+                    b = b.arg_var(proj_idx);
+                    b.add_to(self);
```

Comment:
I was confused at first by how different the `if`/`Ref` and `else`/`AddressOf` cases look here.  But after looking at it more, I think `arg_addr_of(*p)` and `arg_var(dest)` produce the same value, but handle the types differently.  In the `AddressOf` case, `dest` already holds the address of `p` as a raw pointer, so you can just pass `dest` to the instrumentation function.  But in the `Ref` case, `dest` holds the address as a reference, so you take the address as a raw pointer instead using `arg_addr_of(*p)`.  Is that correct?  And would it be legal / reasonable to use `arg_addr_of(*p)` unconditionally?

Reply:
The old code used the reference, but I'm not sure why. `arg_addr_of(*p)` should always be correct.